Tutorials/WinForms/Classic/implement-crud-functionality-with-xpo-objects.md

```diff
@@ -0,0 +1,169 @@
+* Select the `CustomerLayoutControl` component on the design surface.
+* Click the [smart-tag](https://docs.microsoft.com/en-us/dotnet/framework/winforms/controls/performing-common-tasks-using-smart-tags-on-wf-controls) glyph. If the [smart-tag](https://docs.microsoft.com/en-us/dotnet/framework/winforms/controls/performing-common-tasks-using-smart-tags-on-wf-controls) glyph is not visible, press **Esc** several times to navigate from the selected `LayoutGroup` to the `LayoutControl`.
+* Select the **Retrieve Fields** item to open the **Select Binding Source** wizard.
+* Change the **Data Source Update Mode** property value to **OnPropertyChanged** and click the **Next** button.
+* The **Manage Data Bindings** screen displays a list of available properties. Choose the **First Name** and **Last Name** properties, click the **Finish** button, and save the changes.
+* Open the code editor and overload the constructor to accept a nullable integer value (a database record's identifier).
+    ```csharp
+    public EditCustomerForm() {
+        InitializeComponent();
+    }
+    public EditCustomerForm(int? customerId) : this() {
+        CustomerID = customerId;
+    }
+    public int? CustomerID { get; private set; }
+    ```
+* Open the `Form` designer and double click the `Form` header to add the `EditCustomerForm_Load` event handler (you can use the **Properties** window to add it). 
+* Put the following code in the auto-generated `EditCustomerForm_Load` method:
+    ```csharp
+    using DevExpress.Xpo;
+    // ...
+    protected UnitOfWork UnitOfWork { get; private set; }
+    private void EditCustomerForm_Load(object sender, EventArgs e) {
+        UnitOfWork = new UnitOfWork();
+        if (CustomerID.HasValue)
+            CustomerBindingSource.DataSource = UnitOfWork.GetObjectByKey<Customer>(CustomerID.Value);
+        else CustomerBindingSource.DataSource = new Customer(UnitOfWork);
+    }
+    ```
+* Open the `CustomersListForm` designer, select the `CustomersGridView` component in the **Properties** window, and double-click the `RowClick` event in the **Events** section to create the event handler.
+* Add this code to the event handler:
+    ``` csharp
+    using DevExpress.XtraGrid.Views.Grid
+    // ...
+    private void CustomersGridView_RowClick(object sender, RowClickEventArgs e) {
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+            }
+        }
+    }
+    ```
+* Press the **F5** key to run the application, double-click a `GridView` record, and see the result.
+* Close both windows and open the `EditCustomerForm` designer.
+* Select `LayoutControl` and add `SimpleButton` from Toolbox. Add `EmptySpaceItem` elements to adjust the button position and size.
```

Comment:
"from Toolbox"? No article?

Reply:
"adjust the button position and size" Do you mean 'adjust the **button's** position and size'?

Reply:
>"from Toolbox"? No article?

I added the article everywhere.
ff675ffb2bc6a357ae366fe6f75d32354e2458df

>"adjust the button position and size" Do you mean 'adjust the button's position and size'?

Yes, fixed.
951e70deb2a19eee7ee6b3bf0a52bfc1a0e57ced